### DISCUSSION TITLE
Configuration Error during SpaCy Model Training

### DISCUSSION THREAD

**Discussion by RohanPhadnis** [explosion/spaCy#9118] [Help: Installation] [answered]
Hello!

I am working to train a text extraction model using SpaCy. I am using this tutorial: [this tutorial](https://spacy.io/usage/training)

Here is my base_config.cfg file (it's the exact same as in the link):

```
[paths]
train = null
dev = null

[system]
gpu_allocator = null

[nlp]
lang = "en"
pipeline = []
batch_size = 1000

[components]

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH", "SHAPE"]
rows = [5000, 2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"

[training.optimizer]
@optimizers = "Adam.v1"

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001

[initialize]
vectors = ${paths.vectors}
```

I then ran <code>python3 -m spacy init fill-config base_config.cfg config.cfg</code>
That worked flawlessly.

However, when I try <code>python -m spacy train config.cfg --output ./output --paths.train ./train.spacy --paths.dev ./dev.spacy</code>, I get this error: <code>FileNotFoundError: [Errno 2] No such file or directory: 'dev.spacy'</code>

To counter the error, I created an empty placeholder file called <code>dev.spacy</code>
However, that did not work. Here's the error:
<code>ValueError: [E1014] Error loading DocBin data. It doesn't look like the data is in DocBin (.spacy) format. If your data is in spaCy v2's JSON training format, convert it using `python -m spacy convert file.json .`.</code>

I doubt that my data is causing this error. I am using the latest SpaCy and followed every step in the tutorial. Thus, I tried the debug command <code>python3 -m spacy debug data config.cfg</code>
That gave me this error:
<code>ValueError: [E913] Corpus path can't be None. Maybe you forgot to define it in your config.cfg or override it on the CLI?</code>

Here's some additional information that might possibly help:
- **spaCy version:** 3.1.1
- **Platform:** Darwin-20.6.0-x86_64-i386-64bit
- **Python version:** 3.7.7
- **Pipelines:** en_core_web_trf (3.1.0), en_core_web_sm (3.1.0), en_core_web_lg (3.1.0)
(I'm using the en_core_web_lg pipeline for this project).

Would really appreciate any help on this issue. If you need any information about my system, python version, libraries, etc, I can definitely send it over. The only thing I cannot send is my <code>train.spacy</code> file as it contains important data...

Thanks in advance!

**Comment by polm** [comment, marked as answer]
You need to provide an actual data file for `dev.spacy`. If you don't have one you can just use `train.spacy` again (`--paths.dev train.spacy`). But you should hold out some of your data so you can tell your model is actually learning to generalize and not overfitting.

> To counter the error, I created an empty placeholder file called dev.spacy
> I doubt that my data is causing this error.

An empty placeholder file is not a valid DocBin, which is why you're getting that error.

**Reply by RohanPhadnis** [reply]
Thank you so much for the help! How should I initialize the dev.spacy file? I have the DocBin stored in train.spacy

What specifically should the dev.spacy file contain?

I am quite new to the SpaCy framework and don't have much experience. I apologize if my questions are rudimentary...

**Reply by polm** [reply]
The `dev.spacy` file is the same format as the `train.spacy` file, but it's for validation data or the "dev set". 

You should read about the train and dev set in machine learning, this is not specific to spaCy. 

When training your model, in order to make sure it's learning to find patterns in the data and not just memorizing the input, you periodically check it against data you haven't shown it. That's the dev set.

**Reply by RohanPhadnis** [reply]
@polm  Thank you so much! I did not realize that I was lacking a validation dataset. This makes a lot of sense.
